# Data Loading

Import library-library yang akan digunakan

In [47]:
# Import library
import pandas as pd
import numpy as np
import tensorflow as tf

Load dataset dari file csv dan tampilkan dengan pandas dataframe

In [48]:
# Membaca dataset
 
df = pd.read_csv("/content/imdb_top_1000.csv")
df

Poster_Link  \
0    https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1    https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2    https://m.media-amazon.com/images/M/MV5BMTMxNT...   
3    https://m.media-amazon.com/images/M/MV5BMWMwMG...   
4    https://m.media-amazon.com/images/M/MV5BMWU4N2...   
..                                                 ...   
995  https://m.media-amazon.com/images/M/MV5BNGEwMT...   
996  https://m.media-amazon.com/images/M/MV5BODk3Yj...   
997  https://m.media-amazon.com/images/M/MV5BM2U3Yz...   
998  https://m.media-amazon.com/images/M/MV5BZTBmMj...   
999  https://m.media-amazon.com/images/M/MV5BMTY5OD...   

                 Series_Title Released_Year Certificate  Runtime  \
0    The Shawshank Redemption          1994           A  142 min   
1               The Godfather          1972           A  175 min   
2             The Dark Knight          2008          UA  152 min   
3      The Godfather: Part II          1974           A  202 min   
4                12 Angry Men          1957           U   96 min   
..                        ...           ...         ...      ...   
995    Breakfast at Tiffany's          1961           A  115 min   
996                     Giant          1956           G  201 min   
997     From Here to Eternity          1953      Passed  118 min   
998                  Lifeboat          1944         NaN   97 min   
999              The 39 Steps          1935         NaN   86 min   

                        Genre  IMDB_Rating  \
0                       Drama          9.3   
1                Crime, Drama          9.2   
2        Action, Crime, Drama          9.0   
3                Crime, Drama          9.0   
4                Crime, Drama          9.0   
..                        ...          ...   
995    Comedy, Drama, Romance          7.6   
996            Drama, Western          7.6   
997       Drama, Romance, War          7.6   
998                Drama, War          7.6   
999  Crime, Mystery, Thriller          7.6   

                                              Overview  Meta_score  \
0    Two imprisoned men bond over a number of years...        80.0   
1    An organized crime dynasty's aging patriarch t...       100.0   
2    When the menace known as the Joker wreaks havo...        84.0   
3    The early life and career of Vito Corleone in ...        90.0   
4    A jury holdout attempts to prevent a miscarria...        96.0   
..                                                 ...         ...   
995  A young New York socialite becomes interested ...        76.0   
996  Sprawling epic covering the life of a Texas ca...        84.0   
997  In Hawaii in 1941, a private is cruelly punish...        85.0   
998  Several survivors of a torpedoed merchant ship...        78.0   
999  A man in London tries to help a counter-espion...        93.0   

                 Director              Star1              Star2  \
0          Frank Darabont        Tim Robbins     Morgan Freeman   
1    Francis Ford Coppola      Marlon Brando          Al Pacino   
2       Christopher Nolan     Christian Bale       Heath Ledger   
3    Francis Ford Coppola          Al Pacino     Robert De Niro   
4            Sidney Lumet        Henry Fonda        Lee J. Cobb   
..                    ...                ...                ...   
995         Blake Edwards     Audrey Hepburn     George Peppard   
996        George Stevens   Elizabeth Taylor        Rock Hudson   
997        Fred Zinnemann     Burt Lancaster   Montgomery Clift   
998      Alfred Hitchcock  Tallulah Bankhead        John Hodiak   
999      Alfred Hitchcock       Robert Donat  Madeleine Carroll   

              Star3           Star4  No_of_Votes        Gross  
0        Bob Gunton  William Sadler      2343110   28,341,469  
1        James Caan    Diane Keaton      1620367  134,966,411  
2     Aaron Eckhart   Michael Caine      2303232  534,858,444  
3     Robert Duvall    Diane Keaton      1129952   57,300,000  
4     Mart

Gunakan method .info() untuk melihat fitur-fitur pada dataset dan tipe datanya

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Poster_Link    1000 non-null   object 
 1   Series_Title   1000 non-null   object 
 2   Released_Year  1000 non-null   object 
 3   Certificate    899 non-null    object 
 4   Runtime        1000 non-null   object 
 5   Genre          1000 non-null   object 
 6   IMDB_Rating    1000 non-null   float64
 7   Overview       1000 non-null   object 
 8   Meta_score     843 non-null    float64
 9   Director       1000 non-null   object 
 10  Star1          1000 non-null   object 
 11  Star2          1000 non-null   object 
 12  Star3          1000 non-null   object 
 13  Star4          1000 non-null   object 
 14  No_of_Votes    1000 non-null   int64  
 15  Gross          831 non-null    object 
dtypes: float64(2), int64(1), object(13)
memory usage: 125.1+ KB


# Data Understanding dan Data Preparation

Karena sistem rekomendasi ini akan merekomendasikan film berdasarkan genrenya, maka fitur title harus unik. Cek berapa record judul film yang unik.

In [50]:
len(df.Series_Title.unique())

999

Dari 1000 record dataset, ternyata ada 999 record dengan judul yang unik. Cek record mana saja yang judulnya tidak unik.

In [51]:
df.loc[df['Series_Title'].duplicated() == True]

Poster_Link Series_Title  \
136  https://m.media-amazon.com/images/M/MV5BYmJhZm...     Drishyam   

    Released_Year Certificate  Runtime                  Genre  IMDB_Rating  \
136          2015          UA  163 min  Crime, Drama, Mystery          8.2   

                                              Overview  Meta_score  \
136  Desperate measures are taken by a man who trie...         NaN   

            Director       Star1         Star2 Star3         Star4  \
136  Nishikant Kamat  Ajay Devgn  Shriya Saran  Tabu  Rajat Kapoor   

     No_of_Votes    Gross  
136        70367  739,478

Terlihat ada 1 judul film yang redundant.

In [52]:
df.loc[df['Series_Title'] == "Drishyam"]

Poster_Link Series_Title  \
87   https://m.media-amazon.com/images/M/MV5BYmY3Mz...     Drishyam   
136  https://m.media-amazon.com/images/M/MV5BYmJhZm...     Drishyam   

    Released_Year Certificate  Runtime                   Genre  IMDB_Rating  \
87           2013           U  160 min  Crime, Drama, Thriller          8.3   
136          2015          UA  163 min   Crime, Drama, Mystery          8.2   

                                              Overview  Meta_score  \
87   A man goes to extreme lengths to save his fami...         NaN   
136  Desperate measures are taken by a man who trie...         NaN   

            Director       Star1         Star2         Star3         Star4  \
87     Jeethu Joseph    Mohanlal         Meena  Asha Sharath        Ansiba   
136  Nishikant Kamat  Ajay Devgn  Shriya Saran          Tabu  Rajat Kapoor   

     No_of_Votes    Gross  
87         30722      NaN  
136        70367  739,478

Dari fakta di atas, dapat kita simpulkan bahwa film yang redundant ternyata adalah film yang memiliki remake version. Oleh karena itu, judul film tsb harus dipadupadankan dengan tahun rilisnya agar bisa menjadi unik. Kemudian, cek kembali jumlah value unik dari fitur Series_Title

In [53]:
for i in range(len(df.index)):
  if (df['Series_Title'] == "Drishyam")[i] == True:
    df["Series_Title"][i] = df['Series_Title'][i] + ' ' + df['Released_Year'][i]

len(df.Series_Title.unique())

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


1000

Hapus semua fitur, kecuali Series_Title dan Genre, karena keduanya akan menjadi fitur pembentuk sistem rekomendasi.

In [54]:
columns = df.columns.tolist()
wanted_columns = ["Series_Title", "Genre"]

for column in df.columns.tolist():
  if column in wanted_columns:
    columns.remove(column)

df = df.drop(columns=columns)
df

Series_Title                     Genre
0    The Shawshank Redemption                     Drama
1               The Godfather              Crime, Drama
2             The Dark Knight      Action, Crime, Drama
3      The Godfather: Part II              Crime, Drama
4                12 Angry Men              Crime, Drama
..                        ...                       ...
995    Breakfast at Tiffany's    Comedy, Drama, Romance
996                     Giant            Drama, Western
997     From Here to Eternity       Drama, Romance, War
998                  Lifeboat                Drama, War
999              The 39 Steps  Crime, Mystery, Thriller

[1000 rows x 2 columns]

Cek missing value dan hapus jika ada

In [55]:
df.isnull().sum()

Series_Title    0
Genre           0
dtype: int64

# Model Development

Lakukan proses transformasi fitur menggunakan TfidfVectorizer.

In [56]:
from sklearn.feature_extraction.text import TfidfVectorizer
 
# Inisialisasi TfidfVectorizer
tf = TfidfVectorizer()
 
# Melakukan perhitungan idf pada fitur Genre
tf.fit(df['Genre']) 
 
# Mapping array dari fitur index integer ke fitur Series_Title
tf.get_feature_names_out()

array(['action', 'adventure', 'animation', 'biography', 'comedy', 'crime',
       'drama', 'family', 'fantasy', 'fi', 'film', 'history', 'horror',
       'music', 'musical', 'mystery', 'noir', 'romance', 'sci', 'sport',
       'thriller', 'war', 'western'], dtype=object)

In [57]:
# Melakukan fit lalu ditransformasikan ke bentuk matrix
tfidf_matrix = tf.fit_transform(df["Genre"]) 
 
# Melihat ukuran matrix tfidf
tfidf_matrix.shape

(1000, 23)

In [58]:
# Mengubah vektor tf-idf dalam bentuk matriks dengan fungsi todense()
tfidf_matrix.todense()

matrix([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.67835828, 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        ...,
        [0.        , 0.        , 0.        , ..., 0.        , 0.76373063,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.94843765,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.58065827, 0.        ,
         0.        ]])

In [59]:
# Membuat dataframe untuk melihat tf-idf matrix
# Kolom diisi dengan genre
# Baris diisi dengan judul film
 
pd.DataFrame(
    tfidf_matrix.todense(), 
    columns=tf.get_feature_names_out(),
    index=df.Series_Title
).sample(15, axis=1).sample(10, axis=0)

animation     sport  musical   history  \
Series_Title                                                                 
Mississippi Burning                  0.000000  0.000000      0.0  0.801625   
Kari-gurashi no Arietti              0.598369  0.000000      0.0  0.000000   
My Name Is Khan                      0.000000  0.000000      0.0  0.000000   
Moneyball                            0.000000  0.816797      0.0  0.000000   
La montaña sagrada                   0.000000  0.000000      0.0  0.000000   
Bohemian Rhapsody                    0.000000  0.000000      0.0  0.000000   
The Night of the Hunter              0.000000  0.000000      0.0  0.000000   
Ajeossi                              0.000000  0.000000      0.0  0.000000   
The Taking of Pelham One Two Three   0.000000  0.000000      0.0  0.000000   
Paris, Texas                         0.000000  0.000000      0.0  0.000000   

                                      action     drama   fantasy     crime  \
Series_Title                                                                 
Mississippi Burning                 0.000000  0.274262  0.000000  0.531205   
Kari-gurashi no Arietti             0.000000  0.000000  0.000000  0.000000   
My Name Is Khan                     0.000000  1.000000  0.000000  0.000000   
Moneyball                           0.000000  0.219881  0.000000  0.000000   
La montaña sagrada                  0.000000  0.279679  0.783276  0.000000   
Bohemian Rhapsody                   0.000000  0.238507  0.000000  0.000000   
The Night of the Hunter             0.000000  0.175819  0.000000  0.340535   
Ajeossi                             0.678358  0.337068  0.000000  0.652851   
The Taking of Pelham One Two Three  0.560691  0.000000  0.000000  0.539609   
Paris, Texas                        0.000000  1.000000  0.000000  0.000000   

                                        film  horror  comedy  romance  \
Series_Title                                                            
Mississippi Burning                 0.000000     0.0     0.0      0.0   
Kari-gurashi no Arietti             0.000000     0.0     0.0      0.0   
My Name Is Khan                     0.000000     0.0     0.0      0.0   
Moneyball                           0.000000     0.0     0.0      0.0   
La montaña sagrada                  0.000000     0.0     0.0      0.0   
Bohemian Rhapsody                   0.000000     0.0     0.0      0.0   
The Night of the Hunter             0.653117     0.0     0.0      0.0   
Ajeossi                             0.000000     0.0     0.0      0.0   
The Taking of Pelham One Two Three  0.000000     0.0     0.0      0.0   
Paris, Texas                        0.000000     0.0     0.0      0.0   

                                    thriller  biography  family  
Series_Title                                                     
Mississippi Burning                 0.000000   0.000000  0.0000  
Kari-gurashi no Arietti             0.000000   0.000000  0.6628  
My Name Is Khan                     0.000000   0.000000  0.0000  
Moneyball                           0.000000   0.533380  0.0000  
La montaña sagrada                  0.000000   0.000000  0.0000  
Bohemian Rhapsody                   0.000000   0.578562  0.0000  
The Night of the Hunter             0.000000   0.000000  0.0000  
Ajeossi                             0.000000   0.000000  0.0000  
The Taking of Pelham One Two Three  0.628051   0.000000  0.0000  
Paris, Texas                        0.000000   0.000000  0.0000

Gunakan cosine similarity untuk mengukur kemiripan antar film.

In [60]:
from sklearn.metrics.pairwise import cosine_similarity
 
# Menghitung cosine similarity pada matrix tf-idf
cos_sim = cosine_similarity(tfidf_matrix) 
cos_sim

array([[1.        , 0.45876415, 0.33706837, ..., 0.25523548, 0.31696377,
        0.        ],
       [0.45876415, 1.        , 0.73473127, ..., 0.11709289, 0.14541161,
        0.44329292],
       [0.33706837, 0.73473127, 1.        , ..., 0.08603181, 0.10683846,
        0.32570117],
       ...,
       [0.25523548, 0.11709289, 0.08603181, ..., 1.        , 0.80525129,
        0.        ],
       [0.31696377, 0.14541161, 0.10683846, ..., 0.80525129, 1.        ,
        0.        ],
       [0.        , 0.44329292, 0.32570117, ..., 0.        , 0.        ,
        1.        ]])

In [61]:
# Membuat dataframe dari variabel cos_sim dengan baris dan kolom berupa judul film
cos_sim_dataframe = pd.DataFrame(cos_sim, index=df['Series_Title'], columns=df['Series_Title'])
print('Shape:', cos_sim_dataframe.shape)
 
# Melihat similarity matrix untuk tiap film
cos_sim_dataframe.sample(6, axis=1).sample(10, axis=0)

Shape: (1000, 1000)


Series_Title                                Cape Fear  La montaña sagrada  \
Series_Title                                                                
South Park: Bigger, Longer & Uncut           0.000000            0.513534   
Alien                                        0.000000            0.000000   
The Bourne Identity                          0.525594            0.000000   
His Girl Friday                              0.129280            0.089168   
Serbuan maut                                 0.681898            0.000000   
The Broken Circle Breakdown                  0.098082            0.067650   
Le charme discret de la bourgeoisie          0.000000            0.000000   
Dilwale Dulhania Le Jayenge                  0.160325            0.110581   
First Blood                                  0.000000            0.389897   
Star Wars: Episode VI - Return of the Jedi   0.000000            0.828253   

Series_Title                                Sleepers  The Dark Knight  \
Series_Title                                                            
South Park: Bigger, Longer & Uncut          0.000000         0.000000   
Alien                                       0.000000         0.000000   
The Bourne Identity                         0.413352         0.348213   
His Girl Friday                             0.101672         0.107465   
Serbuan maut                                0.536277         0.451767   
The Broken Circle Breakdown                 0.077137         0.081532   
Le charme discret de la bourgeoisie         0.000000         0.000000   
Dilwale Dulhania Le Jayenge                 0.126088         0.133272   
First Blood                                 0.000000         0.482943   
Star Wars: Episode VI - Return of the Jedi  0.000000         0.343080   

Series_Title                                The Warriors      Moon  
Series_Title                                                        
South Park: Bigger, Longer & Uncut              0.000000  0.000000  
Alien                                           0.000000  0.630810  
The Bourne Identity                             0.648933  0.333268  
His Girl Friday                                 0.000000  0.066769  
Serbuan maut                                    0.841916  0.000000  
The Broken Circle Breakdown                     0.000000  0.050656  
Le charme discret de la bourgeoisie             0.000000  0.000000  
Dilwale Dulhania Le Jayenge                     0.000000  0.082803  
First Blood                                     0.399172  0.000000  
Star Wars: Episode VI - Return of the Jedi      0.283570  0.000000

In [62]:
def movie_recommendations(Series_Title, similarity_data=cos_sim_dataframe, features=df[["Series_Title", "Genre"]], n=10):
    """
    Rekomendasi Film berdasarkan kemiripannya
 
    Parameter:
    ---
    Series_Title : tipe data string (str)
                   judul film (index kemiripan dataframe)
    similarity_data : tipe data pd.DataFrame (object)
                      Kesamaan dataframe, simetrik, dengan film sebagai 
                      indeks dan kolom
    features : tipe data pd.DataFrame (object)
               Mengandung fitur-fitur yang digunakan untuk mendefinisikan kemiripan
    n : tipe data integer (int)
        Banyaknya jumlah rekomendasi yang diberikan
    """
 
 
    # Mengambil data dengan menggunakan argpartition untuk melakukan partisi secara tidak langsung sepanjang sumbu yang diberikan    
    # Dataframe diubah menjadi numpy
    # Range(start, stop, step)
    index = similarity_data.loc[:,Series_Title].to_numpy().argpartition(
        range(-1, -n, -1))
    
    # Mengambil data dengan similarity terbesar dari index yang ada
    nearest = similarity_data.columns[index[-1:-(n+2):-1]]
    
    # Drop nama_resto agar nama resto yang dicari tidak muncul dalam daftar rekomendasi
    nearest = nearest.drop(Series_Title, errors='ignore')
 
    return pd.DataFrame(nearest).merge(features).head(n)

# Evaluation

Metrik evaluasi yang relevan untuk sistem rekomendasi seperti ini adalah presisi. Presisi pada proyek ini mengukur banyaknya film hasil rekomendasi yang genrenya relevan dengan genre film masukan. Untuk mempermudah penggunaan berulang, fungsi precision di buat sebagai berikut.

In [63]:
def precision(Series_Title):
  """
    Tingkat presisi sistem rekomendasi
 
    Parameter:
    ---
    Series_Title : tipe data string (str)
                   judul film yang ingin dicari hasil rekomendasinya
    """
  mr_df = movie_recommendations(Series_Title)
  target_genre = df[df["Series_Title"] == Series_Title]["Genre"].tolist()
  target_genre = target_genre[0]
  true_counter = 0
  
  for name in mr_df["Genre"].tolist():
    if name == target_genre:
      true_counter += 1
  
  precision = (true_counter/mr_df.shape[0]) * 100
  print("Tingkat presisi sistem rekomendasi dari film {}: {}%".format(Series_Title, int(precision)))

Mencari film yang mirip dan direkomendasikan setelah film X: First Class, serta menghitung tingkat presisinya.

In [64]:
df[df.Series_Title.eq('X: First Class')]

Series_Title                      Genre
749  X: First Class  Action, Adventure, Sci-Fi

In [65]:
movie_recommendations('X: First Class')

Series_Title                      Genre
0                      Avengers: Infinity War  Action, Adventure, Sci-Fi
1                                   Star Trek  Action, Adventure, Sci-Fi
2                                      Aliens  Action, Adventure, Sci-Fi
3                                   Mad Max 2  Action, Adventure, Sci-Fi
4                                    Predator  Action, Adventure, Sci-Fi
5                                    Serenity  Action, Adventure, Sci-Fi
6             Star Trek II: The Wrath of Khan  Action, Adventure, Sci-Fi
7                            Edge of Tomorrow  Action, Adventure, Sci-Fi
8  Star Wars: Episode VII - The Force Awakens  Action, Adventure, Sci-Fi
9                           The Fifth Element  Action, Adventure, Sci-Fi

In [66]:
precision("X: First Class")

Tingkat presisi sistem rekomendasi dari film X: First Class: 100%


Mencari film yang mirip dan direkomendasikan setelah film The Kid, serta menghitung tingkat presisinya.

In [67]:
df[df.Series_Title.eq('The Kid')]

Series_Title                  Genre
127      The Kid  Comedy, Drama, Family

In [68]:
movie_recommendations('The Kid')

Series_Title                      Genre
0      Miracle on 34th Street      Comedy, Drama, Family
1                 The Sandlot      Comedy, Drama, Family
2            English Vinglish      Comedy, Drama, Family
3                    Amarcord      Comedy, Drama, Family
4                Modern Times      Comedy, Drama, Family
5  The Muppet Christmas Carol      Comedy, Drama, Family
6                  Home Alone             Comedy, Family
7           A Christmas Story             Comedy, Family
8                 The Goonies  Adventure, Comedy, Family
9                      Wonder              Drama, Family

In [69]:
precision("The Kid")

Tingkat presisi sistem rekomendasi dari film The Kid: 60%
